fix(allocator): `Allocator::from_raw_parts` do not use const assertion for endianness test

overlookmotel opened the pull request:
Currently we panic at compile time if attempt to compile `oxc_allocator` with `from_raw_parts` feature enabled on a big-endian system. Instead, panic at runtime.

This will allow `napi/parser` to be built on big endian systems.

The panic will still be optimized out on little-endian systems, and `napi/parser` will not use raw transfer on big endian, so `Allocator::from_raw_parts` won't be called, so this panic should never occur. But the check ensures the soundness of `Allocator::from_raw_parts`.

overlookmotel commented:
* **#10888** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10888?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10888?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10888).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
### Merge activity

* **May 8, 7:42 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(allocator): `Allocator::from_raw_parts` do not use const assertion for endianness test (#10888)

Currently we panic at compile time if attempt to compile `oxc_allocator` with `from_raw_parts` feature enabled on a big-endian system. Instead, panic at runtime.

This will allow `napi/parser` to be built on big endian systems.

The panic will still be optimized out on little-endian systems, and `napi/parser` will not use raw transfer on big endian, so `Allocator::from_raw_parts` won't be called, so this panic should never occur. But the check ensures the soundness of `Allocator::from_raw_parts`.
